redis.go

```diff
@@ -48,83 +48,57 @@ func (hs *hooks) AddHook(hook Hook) {
 func (hs hooks) process(
 	ctx context.Context, cmd Cmder, fn func(context.Context, Cmder) error,
 ) error {
-	ctx, err := hs.beforeProcess(ctx, cmd)
-	if err != nil {
-		cmd.SetErr(err)
-		return err
+	var hookIndex int
+	var retErr error
+
+	for ; hookIndex < len(hs.hooks); hookIndex++ {
```

Comment:
maybe 
```for ; retErr == nil && hookIndex < len(hs.hooks); hookIndex++```
looks nice, 
and  idx-- is not required at afterprocess starts.

Reply:
I believe with `retErr == nil` I need `idx -= 2` if I don't want to include the hook that returned the error. Am I missing something?

Reply:
example:   
beforeProcess is stop at i'th idx (i>=0),  afterProcess should starts at hooks[i], not hooks[i-1]

Reply:
code:
```golang
// after hook
for hookIndex--; hookIndex >= 0; hookIndex--
```
hookIndex = i
the  first `hookIndex--` will be set i-1.  it will lose the i'th hook.

Reply:
Yes, I intentionally skipped the AfterProcess hook that returned the error from BeforeProcess. Since I don't have strong preference I've changed the code to include it.

Please take another look.

Reply:
if beforehooks run all success, hookIndex is equals to len(hooks),
afterhooks will be run out of range.

Reply:
there is example,  register a hook wicth feature is a authentification feature,  
if error is returns from this beforehook,  

the i'th afterhook  should record the error and emit it's metrics  with error msg.

Reply:
example:  
there is a hook registered in hooks at i'th index, 
beforehook is focus on authentification, 
afterhook records log and emit metrics,  the error msg is used in afterhook.